colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py

```diff
@@ -76,6 +77,40 @@ def __init__(
         self.verbose = verbose
         self.coordinator = DistCoordinator()
 
+    @staticmethod
+    def _named_modules(
+        module: nn.Module, memo: Optional[Set[nn.Module]] = None, prefix: str = "", remove_duplicate: bool = True
+    ):
+        r"""Returns an iterator over all leaf modules in the network, yielding
+        both the name of the module as well as the module itself.
+
+        Args:
+            memo: a memo to store the set of modules already added to the result
+            prefix: a prefix that will be added to the name of the module
+            remove_duplicate: whether to remove the duplicated module instances in the result
+                or not
+
+        Yields:
+            (str, Module): Tuple of name and module
+
+        Note:
+            Duplicate modules are returned only once. In the following
+            example, ``l`` will be returned only once.
+        """
+        if memo is None:
+            memo = set()
+
+        if module not in memo:
+            sub_modules = [(name, subm) for (name, subm) in module._modules.items() if subm is not None]
+            if len(sub_modules) == 0:
+                if remove_duplicate:
+                    memo.add(module)
+                yield prefix, module
+            else:
+                for name, subm in sub_modules:
+                    submodule_prefix = prefix + ("." if prefix else "") + name
+                    yield from HybridParallelCheckpointIO._named_modules(subm, memo, submodule_prefix, remove_duplicate)
```

Comment:
This cannot handle such networks:
![image](https://github.com/hpcaitech/ColossalAI/assets/23111350/7e0011f5-5251-45af-a3a7-fabfb362e4f1)
![image](https://github.com/hpcaitech/ColossalAI/assets/23111350/5a55908f-45e4-402e-bb5c-c983685c998d)

Then `model.p` is missing here